File: backend/src/main/java/com/whiteblue/backend/security/jwt/JWTProvider.java
package com.whiteblue.backend.security.jwt;

import com.whiteblue.backend.config.ApplicationProperties;
import com.whiteblue.backend.domain.user.UserRepository;
import com.whiteblue.backend.security.oAuth.OAuthUser;
import io.jsonwebtoken.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import javax.annotation.PostConstruct;
import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import java.util.stream.Collectors;

@Log4j2
@RequiredArgsConstructor
@Component
public class JWTProvider {
    private final ApplicationProperties applicationProperties;

    private final UserRepository userRepository;

    private Long accessTokenExpire;

    private Long refreshTokenExpire;

    private String secretKey;

    @PostConstruct
    public void init() {
        final Long HOUR = 1000L * 60 * 60;
        accessTokenExpire = applicationProperties.getAccessTokenExpire() * HOUR;
        refreshTokenExpire = applicationProperties.getRefreshTokenExpire() * HOUR;
        secretKey = applicationProperties.getSecretKey();
    }

    public String createAccessToken(Authentication authentication) {
        Date now = new Date();
        OAuthUser user = (OAuthUser) authentication.getPrincipal();

        return Jwts.builder()
                   .setSubject("access_token")
                   .setIssuer("Whiteblue")
                   .setIssuedAt(now)
                   .claim("id", user.getId())
                   .claim("username", user.getUsername())
                   .claim("name", user.getName())
                   .claim("role", authentication.getAuthorities()
                                                .stream()
                                                .map(GrantedAuthority::getAuthority)
                                                .collect(Collectors.joining(",")))
                   .setExpiration(new Date(now.getTime() + accessTokenExpire))
                   .signWith(SignatureAlgorithm.HS512, secretKey)
                   .compact();
    }

    @Transactional
    public String createRefreshToken(Authentication authentication) {
        Date now = new Date();
        String refreshToken = Jwts.builder()
                                  .setSubject("refresh_token")
                                  .setIssuer("Whiteblue")
                                  .setIssuedAt(now)
                                  .setExpiration(new Date(now.getTime() + refreshTokenExpire))
                                  .signWith(SignatureAlgorithm.HS512, secretKey)
                                  .compact();
        OAuthUser user = (OAuthUser) authentication.getPrincipal();
        userRepository.updateRefreshToken(user.getId(), refreshToken);

        return refreshToken;
    }

    public Authentication getAuthentication(String accessToken) {
        Claims claims = parseClaims(accessToken);
        Collection<? extends GrantedAuthority> authorities =
                Arrays.stream(claims.get("role")
                                    .toString()
                                    .split(","))
                      .map(SimpleGrantedAuthority::new)
                      .collect(Collectors.toList());

        OAuthUser principal = OAuthUser.builder()
                                       .id(((Integer) claims.get("id")).longValue())
                                       .username((String) claims.get("username"))
                                       .name((String) claims.get("name"))
                                       .authorities(authorities)
                                       .build();

        return new UsernamePasswordAuthenticationToken(principal, "", authorities);
    }

    public Boolean validateToken(String token) {
        try {
            Jwts.parser()
                .setSigningKey(secretKey)
                .parseClaimsJws(token);
            return true;
        } catch (ExpiredJwtException exception) {
            log.info("만료된 토큰입니다.");
        } catch (UnsupportedJwtException exception) {
            log.info("지원되지 않는 토큰입니다.");
        } catch (IllegalStateException exception) {
            log.info("토큰이 잘못되었습니다");
        } catch (SignatureException exception) {
            log.info("토큰 서명이 적절하지 않습니다.");
        }
        return false;
    }

    public Claims parseClaims(String accessToken) {
        try {
            return Jwts.parser()
                       .setSigningKey(secretKey)
                       .parseClaimsJws(accessToken)
                       .getBody();
        } catch (ExpiredJwtException exception) {
            return exception.getClaims();
        }
    }
}

File: backend/src/main/kotlin/com/whiteblue/backend/security/jwt/JWTProvider.kt
package com.whiteblue.backend.security.jwt

import com.whiteblue.backend.config.ApplicationProperties
import com.whiteblue.backend.domain.user.UserRepository
import com.whiteblue.backend.security.oAuth.OAuthUser
import io.jsonwebtoken.*
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Component
class JWTProvider(
    private val applicationProperties: ApplicationProperties,

    private val userRepository: UserRepository
) {
    companion object {
        const val HOUR = 1000L * 60 * 60
    }

    private val secretKey = applicationProperties.secretKey

    private val accessTokenExpire = applicationProperties.accessTokenExpire * HOUR

    private val refreshTokenExpire = applicationProperties.refreshTokenExpire * HOUR

    fun createAccessToken(authentication: Authentication): String {
        val now = Date()
        val user = authentication.principal as OAuthUser

        return Jwts.builder()
            .setSubject("access_token")
            .setIssuer("Whiteblue")
            .setIssuedAt(now)
            .claim("id", user.id)
            .claim("username", user.username)
            .claim("name", user.name)
            .claim("role", authentication.authorities.joinToString(",") { it.authority })
            .setExpiration(Date(now.time + accessTokenExpire))
            .signWith(SignatureAlgorithm.HS512, secretKey)
            .compact()
    }

    @Transactional
    fun createRefreshToken(authentication: Authentication): String {
        val now = Date()
        val refreshToken = Jwts.builder()
            .setSubject("refresh_token")
            .setIssuer("Whiteblue")
            .setIssuedAt(now)
            .setExpiration(Date(now.time + refreshTokenExpire))
            .signWith(SignatureAlgorithm.HS512, secretKey)
            .compact()
        val user = authentication.principal as OAuthUser

        userRepository.updateRefreshToken(user.id, refreshToken)

        return refreshToken
    }

    fun getAuthentication(accessToken: String): Authentication {
        val claims = try {
            Jwts.parser()
                .setSigningKey(secretKey)
                .parseClaimsJws(accessToken)
                .body
        } catch (exception: ExpiredJwtException) {
            exception.claims
        }
        val authorities = claims["role"]
            .toString()
            .split(",")
            .map(::SimpleGrantedAuthority)
        val principal = OAuthUser(
            id = (claims["id"] as Int).toLong(),
            username = claims["username"] as String,
            name = claims["name"] as String,
            authorities = authorities,
            attributes = null
        )

        return UsernamePasswordAuthenticationToken(principal, "", authorities)
    }

    fun validateToken(token: String): Boolean {
        try {
            Jwts.parser()
                .setSigningKey(secretKey)
                .parseClaimsJws(token)
            return true
        } catch (exception: ExpiredJwtException) {
            println("만료된 토크입니다")
        } catch (exception: UnsupportedJwtException) {
            println("지원되지 않는 토크입니다")
        } catch (exception: IllegalStateException) {
            println("토큰이 잘못되었습니다")
        } catch (exception: SignatureException) {
            println("토큰 서명이 적절하지 않습니다.")
        }
        return false
    }
}
